### PR TITLE
ja-slide skill: Round 12/13 fixes + slide preview UI overhaul

### PR DESCRIPTION
**Pull request by Rekez886**
This branch lands the ja-slide skill iteration plus the slide preview UI work that supports it.

## Skill changes (`skills/ja-slide/`)

- **Round 12** — added per-canonical `title_fit: { min, ideal, max, lines }` to `layouts.json` so layout selection rejects canonicals where the section title would either wrap (too long) or leave the placeholder visually empty (too short). Added Step 3.7.0 to SKILL.md with the cells-based filtering algorithm and a two-symptom self-check row.

- **Round 13: image-marker priority** — when a section has both a scenario hint (e.g. `kpi_summary` → T+3B) AND an `[image:]` marker, the agent must pick T+IMG. Earlier rounds drifted into T+3B and silently dropped the image. Inlined the override into every scenario hint in `layouts.json` and added a callout above SKILL.md §3.1 so the rule is unmissable from either source of truth.

- **Round 13 visual-review followups** —
  - T+3B title sits mid-bottom (above the 3-arrow row); flagged with a `title_position` warning so headline-emphasis sections route to T+SUB / T-ONLY instead.
  - T-LONG quotes leave 1-mora particles alone at line start at large font; added a self-check row + §5.1 in prompt-rules.md.
  - T+IMG (P31) explicit `\n` demotes line 2+ to paragraph style; SKILL.md §9.5.3 stage 1.5 spells out the priority (rewrite to one line, then accept the demotion + warn user, then switch to T+SUB).
  - SKILL.md §8.1.1 makes the `PATCH /api/projects/:id` call mandatory right after deck copy — without it the daemon's `/thumbnails`, `/deck-pdf`, `/page-ids`, `/page-image` endpoints all 404.

## Web changes (`apps/web/`)

- Slide preview now renders the deck PDF on a canvas via pdfjs (no browser PDF chrome / gray padding).
- Rebrand pass acted on a design critique: card meta dropped skill-id and broken status enum, sub-tabs that always returned the same list removed, sidebar form scoped to a single design system, sidebar title swapped from editorial serif to system sans, project header humanizes ISO datestamps and stops exposing skill ids, chat starter prompts rewritten for slide use cases, viewer toolbar shows page count + clearer "Edit in Google Slides" CTA, pager uses chevron icons and adds the ページ unit.
- Deck preview's two image-warning panels deduped (newer `imageSlots` wins; legacy `missingImages` is now a fallback only).

## Commit history

History was rewritten to replace internal project codenames and reviewer names with placeholder tokens (`<reviewer>`, `<colleague>`, `<newsletter-project>`, `<brand-project>`, `<industry-event>`, `<hq-design-lead>`) in both the diff text and commit messages. Verified zero hits for the original tokens before opening this PR.

## Test plan

- [x] Skill: Round 13 stress test on a 10-page deck with KPI + image-marker section reproduced as T+IMG (no image drop). Title fit pass on all 10 sections.
- [x] Web: `pnpm --filter @open-design/web typecheck`
- [ ] Manual: open a project's deck preview, confirm thumbnails load, only one image-warning panel renders.

### CODE REVIEW

**Comment by lefarcen** [comment]
Hi @Rekez886! 🎉
Thanks for the contribution — this lands both the wix-ja-slide skill iteration and the slide preview UI improvements.
I will run a deep review and get back to you within 24h.

Thanks for making open-design better!
— open-design team

**Comment by mrcfps** [comment]
@Rekez886 I couldn't generate review comments for nexu-io/open-design#568 because this pull request currently has merge conflicts.

Please resolve the conflicts with main, push the updated branch, and then request or wait for the review to run again.




<sub>Generated by Looper 0.5.6 · runner=reviewer · agent=opencode</sub>

**Review by lefarcen — COMMENTED** [review]
Hi @Rekez886 — this is a substantial piece of work that brings Google Slides integration + the wix-ja-slide skill to life. I've reviewed both the code and the reasoning/design, and overall this is **production-ready** with a few observations below (all non-blocking).

## Lens A — Code correctness

**Security & auth handling:**
- ✅ OAuth token storage uses mode 0600 and lives in `~/.open-design/`, appropriate for a desktop daemon
- ✅ `uploadImage` path traversal guard: `resolved.startsWith(root + path.sep)` correctly prevents `../../` escapes
- ✅ Drive sharing model: best-effort public grant with DLP-aware silent fallback is the right call for Wix Workspace environments
- ✅ `buildRedirectUri` uses the request's own host so OAuth works across dynamic daemon ports

**Boundary conditions & input validation:**
- ✅ `duplicateSlide` preflight: validates idMap destination format ([A-Za-z0-9_-]+, 5-50 chars), checks source key existence, collision detection — thorough
- ✅ `updateSlidesPosition` one-at-a-time move strategy avoids the Google API's in-order constraint violation
- ✅ Error wrapping with `.code` fields (`GOOGLE_AUTH_REQUIRED`, `GOOGLE_API_ERROR`, etc.) gives route handlers stable HTTP status mapping

**Maintainability & architecture:**
- ✅ `updateTextByObjectId` per-paragraph style preservation: captures original runs' styles, applies by index to new paragraphs, fallback to first-run style if batch fails — handles the P31 mixed-style placeholder correctly
- ✅ Thumbnail cache (25min TTL) reduces Google API spam when the Designs grid re-renders
- ✅ `deletePages` chunking (200 requests/batch) stays well under Google's 1000-req cap

**Testing gap (non-blocking):**
- No vitest unit tests for `google-auth.ts` / `google-slides.ts` / daemon routes. The derisk script (`test-google-auth.ts`) is a good manual smoke test, but automated coverage for auth refresh, token expiry, collision edge cases would future-proof this. Consider adding in a follow-up.

## Lens B — Reasoning completeness (SKILL.md, DESIGN.md)

**skills/wix-ja-slide/SKILL.md:**
- ✅ Assumptions explicit: "Full-width JP chars ≈ 2× latin width" motivates title_fit filtering; "T+IMG explicit `\n` demotes line 2+ to paragraph style" warns of the rendering quirk
- ✅ Failure modes covered: §5.1 warns of 1-mora particles left at line start (large font + T-LONG); §8.1.1 makes `PATCH /api/projects/:id` mandatory (prevents 404 on thumbnails/pdf)
- ✅ Open questions surfaced: Round 13 image-marker priority was a "drift" correction — implies earlier iterations silently dropped images, now fixed by inlining override into every scenario hint

**design-systems/wix-japan/DESIGN.md:**
- ✅ Rebrand rationale: "design critique: card meta dropped skill-id and broken status enum, sub-tabs that always returned the same list removed" — concrete trigger for the changes
- ✅ Consistency with repo philosophy: locales pruned to en/ja/zh-CN aligns with "local-first / open-source" (less translation debt for a niche skill)

**Minor observation (non-blocking):**
- Commit history rewrite (placeholder tokens for `<reviewer>`, `<colleague>`, etc.) is thorough, but the PR body could briefly note *why* history was rewritten (e.g., "removed internal codenames before open-sourcing") for future archaeology.

## Overall

Looks good to me; deferring final approval to a maintainer. The Google API integration is solid, the visual self-check loop (PDF export → Read pages → regenerate if overflow) is the right design, and the Wix Japan rebrand pass is clean. No must-fix blockers.

Let me know if you'd like me to clarify any of the above!

**Comment by mrcfps** [comment]
@Rekez886 I couldn't generate review comments for nexu-io/open-design#568 because this pull request currently has merge conflicts.

Please resolve the conflicts with main, push the updated branch, and then request or wait for the review to run again.




<sub>Generated by Looper 0.5.6 · runner=reviewer · agent=opencode</sub>

**Comment by Rekez886** [comment]
Ran `codex review --base main` against this branch as a second-opinion pass.

**Codex findings (2):**

- **[P1] Silent private-Drive fallback breaks image insertion under Workspace DLP** — `apps/daemon/src/google-slides.ts:176-180`. Pre-existing code (not touched by this PR), but Codex flagged it as a real bug that affects the Google Workspace flow this skill targets. Filed as follow-up: #569. **Out of scope for this PR.**

- **[P2] `apps/web/public/pdf.worker.min.mjs` fails `pnpm check:residual-js`** — directly caused by this PR's pdfjs canvas commit (`8dc6a55`). **Fixed in `e5f33ac`** by adding the worker bundle to the documented exact-path allowlist next to other vendored runtime assets. Verified `pnpm check:residual-js` passes.

Codex did not flag the rest of the PR's surface area (SKILL.md rule additions §3.0.6 / §3.7.0 / §8.1.1 / §9.5, layouts.json title_fit & title_position fields, image-warning panel dedupe, history scrub).

Updated branch tip: `e5f33ac`.

**Comment by Rekez886** [comment]
Merged `upstream/main` into the branch (commit `2bfd5ef`) to resolve the conflicts that piled up since the branch was forked.

**Conflict resolution summary** (full list in the merge commit message):

- Restored the 9 i18n locales an earlier rebrand commit had wrongly deleted — `ja-slide` is a single skill in a multi-locale OSS project, so removing locales here was out of scope.
- Migrated the `apps/web/public/pdf.worker.min.mjs` allowlist entry from the deleted `scripts/check-residual-js.ts` to the new `scripts/guard.ts`.
- Took upstream wholesale for `DesignsTab.tsx` and `EntryView.tsx` — the team-specific rebrand simplifications there are deployment-specific and incompatible with upstream's connector + live-artifact additions.
- Lazy-imported `pdfjs-dist` inside `PdfCanvasPage` so jsdom-based tests don't trip over `DOMMatrix`.
- Combined `FileWorkspace.tsx`: live-artifact title plumbing (upstream) + `result.json` → "プレビュー" rename (this branch) coexist.
- Re-added the eight new i18n keys (`workspace.deckPreview`, `fileViewer.googleSlides*`) to `fa`, `pt-BR`, `ru`, `uk`, `zh-TW` with English-stub values so the `Dict` type is satisfied. Added the JA design system to FR/RU design-system fallback + `ja-slide` to FR/RU skill fallback so `content.test.ts` passes.

**Validation locally:**

- `pnpm guard` ✅
- `pnpm --filter @open-design/web typecheck` ✅
- `pnpm --filter @open-design/web test` ✅ 241 / 241
- `pnpm --filter @open-design/daemon test` ✅ 761 / 761

PR is now `MERGEABLE`. Ready for review.

**Review by lefarcen — COMMENTED** [review]
Conflicts resolved — merge commit `2bfd5ef` looks clean:

✅ Restored 9 locales wrongly deleted in rebrand (correct — `wix-ja-slide` shouldn't remove unrelated i18n)
✅ Migrated `pdf.worker.min.mjs` allowlist to new `guard.ts` (maintains the P2 fix from earlier)
✅ Lazy-import `pdfjs-dist` in `PdfCanvasPage` (standard jsdom workaround)
✅ i18n stub additions (fa/pt-BR/ru/uk/zh-TW) maintain type safety
✅ Local validation passed (guard + typecheck + web 241/241 + daemon 761/761)

Codex verified the core skill/web changes earlier — no issues beyond the pre-existing P1 (filed as #569, out of scope). The merge commit integrates upstream without introducing new surface area.

**Ready from my side** — deferring final approval to a maintainer.

File: apps/daemon/src/server.ts [hunk 1]
```diff
@@ -3947,6 +4061,352 @@ export async function startServer({ port = 7456, host = process.env.OD_BIND_HOST
...
+  });
+
+  app.get('/api/google/auth/callback', async (req, res) => {
+    const error = typeof req.query.error === 'string' ? req.query.error : null;
+    const code = typeof req.query.code === 'string' ? req.query.code : null;
+    if (error) {
+      res
+        .status(400)
+        .type('html')
+        .send(`<h1>Google OAuth error</h1><p>${error}</p><p>You can close this tab.</p>`);
```
Comment (mrcfps) [inline apps/daemon/src/server.ts]: The OAuth callback error path renders the raw `error` query parameter directly into an HTML response. A request such as `/api/google/auth/callback?error=<script>...</script>` would execute attacker-controlled script on the daemon origin, which is especially risky because this localhost origin also hosts the privileged `/api/*` endpoints. Please either return plain text/JSON for this branch or HTML-escape all user-controlled values before interpolating them into the response.

File: apps/daemon/src/google-slides.ts [hunk 2]
```diff
@@ -0,0 +1,926 @@
...
+    return {
+      ok: true,
+      preservedParagraphStyles: paragraphStyles.length,
+      newParagraphs: newParagraphs.length,
+    };
+  } catch (err) {
+    if (hadPriorText) {
+      try {
+        const fallbackRequests = [
+          { insertText: { objectId, text, insertionIndex: 0 } },
```
Comment (mrcfps) [inline apps/daemon/src/google-slides.ts:430]: If the styled replacement batch fails, this fallback inserts the new text at index 0 without first deleting the existing placeholder text. Slides `batchUpdate` is atomic, so the failed first batch leaves the old text in place; this fallback then prepends the replacement and produces duplicated/garbled slide copy instead of a true replacement. Please include the same `deleteText` request before `insertText` in the fallback path (or otherwise ensure the old text is cleared) before applying the simplified style update.

**Review by mrcfps — CHANGES_REQUESTED** [review]
@Rekez886 thanks for the substantial wix-ja-slide and slide-preview work here — the skill guidance and UI path are moving in a useful direction. I found a few security/correctness issues in changed code that should be fixed before merge.



<sub>Generated by Looper 0.5.6 · runner=reviewer · agent=opencode</sub>

Location: apps/web/src/components/FileViewer.tsx RIGHT line 4070

`deckUrl` comes directly from the generated `result.json` and is rendered as an anchor `href` without the same allowlist validation used for Slides embeds. Any artifact writer that can produce `result.json` can set `deckUrl` to `javascript:...` or a non-Google phishing URL, and the toolbar will present it as the trusted “Edit in Google Slides” action. Please derive the edit URL from a validated deck ID, or validate `deckUrl` against a strict `https://docs.google.com/presentation/d/<id>/...` allowlist before rendering the link.



Inline comment could not be anchored: inline anchor is outside the PR diff anchorable ranges

File: apps/daemon/src/server.ts (hunk 1 above)
Comment (mrcfps) [inline apps/daemon/src/server.ts]: The OAuth callback error branch interpolates the raw `error` query parameter into an HTML response. For example, `/api/google/auth/callback?error=<script>...</script>` would execute on the daemon origin, and that same localhost origin hosts privileged `/api/*` endpoints. The catch path below has the same pattern with `err.message` inside `<pre>`. Please either return plain text/JSON for these error responses or HTML-escape every interpolated value before sending HTML.

File: apps/daemon/src/google-slides.ts (hunk 2 above)
Comment (mrcfps) [inline apps/daemon/src/google-slides.ts:430]: If the styled replacement batch fails, this fallback inserts the new text at index 0 without clearing the existing shape text first. Slides `batchUpdate` is atomic, so the failed main batch leaves the original placeholder content untouched; this fallback then prepends the replacement and produces duplicated/garbled slide copy instead of a replacement. Please include the same `deleteText` request before this fallback `insertText` (or otherwise clear the old text) before applying the simplified style update.

**Review by mrcfps — CHANGES_REQUESTED** [review]
@Rekez886 thanks for the continued Wix Japan slide skill and preview work here — the skill iteration notes are thoughtful, and the PDF canvas preview path is a useful direction. I found a few correctness/security issues in changed ranges that should be fixed before merge, including one CSS parse blocker and some unresolved daemon/viewer safety issues on the current head.



<sub>Generated by Looper 0.6.0 · runner=reviewer · agent=opencode</sub>

Location: apps/web/src/index.css RIGHT line 9084

This new `.google-slides-warnings li` rule never closes before the Pet section starts. Because the added line ends with `line-height: 1.5;` and no `}`, the browser keeps parsing the following `.pet-*` CSS inside an unterminated rule, so the rest of the stylesheet after this point can be dropped or misparsed. That breaks unrelated UI surfaces (the Pet overlay/settings styles and anything later in `index.css`) whenever this stylesheet is loaded. Please add the missing closing brace immediately after this declaration and consider running a CSS parser/formatter check so this class of regression is caught automatically.



Inline comment could not be anchored: inline anchor is outside the PR diff anchorable ranges

Location: apps/web/src/components/FileViewer.tsx RIGHT line 4070

`deckUrl` comes directly from the generated `result.json` and is rendered as an anchor `href` without the allowlist validation used for Slides embed URLs. Any artifact writer that can produce this manifest can set `deckUrl` to a `javascript:` URL or a non-Google phishing URL, and the toolbar presents it as the trusted “Edit in Google Slides” action. Please derive the edit URL from a validated `deckId`, or validate `deckUrl` against a strict `https://docs.google.com/presentation/d/<id>/...` allowlist before rendering the link.



Inline comment could not be anchored: inline anchor is outside the PR diff anchorable ranges

**Comment by Rekez886** [comment]
Addressed all 3 P1 + 2 P2 findings from the Codex and Looper opencode reviews in commit `203c787`:

## P1 (security)

| # | File | Finding | Fix |
|---|---|---|---|
| 1 | `apps/daemon/src/server.ts` | OAuth callback raw-HTML interpolation (XSS via `?error=<script>`) | Inline `esc()` HTML-escapes both `error` query and `err.message` before sending |
| 2 | `apps/web/src/components/FileViewer.tsx` | `deckUrl` from result.json rendered as `<a href>` without validation (`javascript:` / phishing URL injection) | Switched to deriving the link from `result.deckId` (`/^[A-Za-z0-9_-]{10,}$/`) and building `https://docs.google.com/presentation/d/<id>/edit` client-side. `result.deckUrl` is no longer read |
| 3 | `apps/web/src/index.css` | `.google-slides-warnings li` rule missing `}` → browser swallows following `.pet-*` styles | Added closing brace |

## P2 (correctness)

| # | File | Finding | Fix |
|---|---|---|---|
| 4 | `apps/daemon/src/google-slides.ts` | `update-text` fallback inserts at index 0 without first deleting → atomic batch failure leaves old + new text concatenated | Bundled `deleteText { type: 'ALL' }` + `insertText` + style into single atomic batch; renamed fallback flag to `delete-insert` |
| 5 | `apps/daemon/src/google-auth.ts` | Token refresh merges into the boot-time snapshot, clobbering later credential rotations | Merge base switched to `client.credentials` (live state) for both `getAuthClient` and `exchangeCode` listeners |

## Validation

- `pnpm guard` ✅
- `pnpm --filter @open-design/web typecheck` ✅
- `pnpm --filter @open-design/daemon typecheck` ✅
- `pnpm --filter @open-design/web test` ✅ 241/241
- `pnpm --filter @open-design/daemon test` ✅ 761/761

The pre-existing P1 about Drive DLP silent fallback (`google-slides.ts:176-180`) tracked in #569 is intentionally left for follow-up — it's not in this branch's diff and the fix touches the auth path semantics, which feels worth a separate PR.

Ready for re-review.

**Review by lefarcen — COMMENTED** [review]
All 4 P1 findings from mrcfps's review are fixed in commit `203c787`:

✅ **CSS parse blocker** — `.google-slides-warnings li` now closes properly (line 9085)
✅ **`deckUrl` XSS** — switched to validated `result.deckId` (`/^[A-Za-z0-9_-]{10,}$/`) + derived `https://docs.google.com/presentation/d/<id>/edit` URL; no longer trusts raw `result.deckUrl`
✅ **OAuth callback XSS** — both `error` query parameter and `err.message` now HTML-escaped via inline `esc()` helper
✅ **Slides fallback duplication** — fallback bundles `deleteText { type: 'ALL' }` + `insertText` into single atomic batch

The token refresh merge fix (P2 #5 in your comment) also looks correct — `getAuthClient` and `exchangeCode` listeners now merge into `client.credentials` (live state) instead of boot-time snapshot.

**Ready from my side** — deferring final approval to a maintainer.

**Comment by mrcfps** [comment]
@Rekez886 I'm holding off on generating review comments for nexu-io/open-design#568 because this pull request has merge conflicts right now.

Please resolve the conflicts with main and push the updated branch. Once that's done, request or wait for the review to run again and I'll take another look.




<sub>Generated by Looper 0.6.0 · runner=reviewer · agent=codex</sub>

File: apps/daemon/src/server.ts [hunk 3]
```diff
@@ -3961,6 +4075,369 @@ export async function startServer({ port = 7456, host = process.env.OD_BIND_HOST
...
+        );
+    } catch (err) {
+      res
+        .status(500)
+        .type('html')
+        .send(`<h1>OAuth exchange failed</h1><pre>${esc(err.message)}</pre>`);
+    }
+  });
+
+  app.post('/api/google/auth/clear', async (_req, res) => {
```
Comment (mrcfps) [inline apps/daemon/src/server.ts:4170]: The new Google OAuth/Slides routes are registered without the existing `requireLocalDaemonRequest` guard used for other sensitive daemon surfaces. That leaves at least `POST /api/google/auth/clear` reachable as a cross-site form POST from any website a user visits (modern browsers include an `Origin`, but nothing checks it here), and the same unguarded route group later exposes Drive/Slides mutation endpoints once JSON requests come from a permitted local UI. Because these routes hold broad Google Drive credentials and can mutate decks, please put the `/api/google/*` router behind `requireLocalDaemonRequest` (or apply the same loopback Host/Origin validation to each handler; the OAuth callback should still work because top-level Google redirects arrive from loopback with no hostile Origin).

**Review by mrcfps — CHANGES_REQUESTED** [review]
@Rekez886 thanks for the careful follow-up fixes and the continued wix-ja-slide polish here — the previous XSS/deck URL/CSS/token issues look addressed on this head. I found one remaining local-daemon protection gap around the newly added Google routes that should be fixed before merge.



<sub>Generated by Looper 0.6.1 · runner=reviewer · agent=opencode</sub>

**Review by abyisysfogoso-stack — APPROVED** [review]
